File: MISA.AssetManagement.Webapp/MISA.AssetManagement.Fresher.Webapp/src/domains/models/FixedAsset.js
/**
 * Lớp mô hình đại diện cho Tài sản cố định (FixedAsset)
 * Dùng để ánh xạ dữ liệu từ API và chuẩn hóa dữ liệu hiển thị trong UI.
 * CreatedBy: TTVinh (17/11/2025)
 * UpdatedBy: TTVinh (18/11/2025)
 */
export class FixedAsset {
  /**
   * @constructor
   * @param {Object} [data={}] - Dữ liệu khởi tạo tài sản (nhận từ API hoặc form)
   */
  constructor(data = {}) {
// #region Thuộc tính cơ bản
    this.FixedAssetId = data.FixedAssetId || data.fixedAssetId || null
    this.FixedAssetCode = data.FixedAssetCode || data.fixedAssetCode || ''
    this.FixedAssetName = data.FixedAssetName || data.fixedAssetName || ''
    // #endregion

    // #region Thông tin bộ phận sử dụng
    this.DepartmentId = data.DepartmentId || data.departmentId || null
    this.DepartmentCode = data.DepartmentCode || data.departmentCode || ''
    this.DepartmentName = data.DepartmentName || data.departmentName || ''
    // #endregion

    // #region Thông tin loại tài sản
    this.FixedAssetCategoryId = data.FixedAssetCategoryId || data.fixedAssetCategoryId || null
    this.FixedAssetCategoryCode = data.FixedAssetCategoryCode || data.fixedAssetCategoryCode || ''
    this.FixedAssetCategoryName = data.FixedAssetCategoryName || data.fixedAssetCategoryName || ''
    // #endregion

    // #region Thông tin giá trị, hao mòn
    this.Quantity = data.Quantity || data.quantity || 0
    this.Cost = data.Cost || data.cost || 0
    this.LifeTime = data.LifeTime || data.lifeTime || 0
    this.DepreciationRate = data.DepreciationRate || data.depreciationRate || 0
    this.DepreciationValue = data.DepreciationValue || data.depreciationValue || 0
    this.AccumulatedDepreciation = data.DepreciationValue || data.depreciationValue || 0
    this.RemainingValue = data.Cost - data.DepreciationValue || data.cost - data.depreciationValue || 0
    // #endregion

    // #region Thông tin thời gian
    this.PurchaseDate = data.PurchaseDate || data.purchaseDate || ''
    this.StartUsingDate = data.PurchaseDate || data.purchaseDate || ''
    this.ProductionYear = data.ProductionYear || data.productionYear || 0
    this.TrackedYear = data.TrackedYear || data.trackedYear || 0
    // #endregion

    // #region Mô tả
    this.Description = data.Description || data.description || ''
    this.IsActive = data.IsActive || data.isActive || true
    this.CreatedDate = data.CreatedDate || data.createdDate || ''
    this.CreatedBy = data.CreatedBy || data.createdBy || null
    this.ModifiedDate = data.ModifiedDate || data.modifiedDate || ''
    this.ModifiedBy = data.ModifiedBy || data.modifiedBy || null
    // #endregion
  }

  // #region Getters

  /** @returns {string|null} ID tài sản */
  get id() {
    return this.FixedAssetId
  }

  /** @returns {string} Mã tài sản */
  get assetCode() {
    return this.FixedAssetCode
  }

  /** @returns {string} Tên tài sản */
  get assetName() {
    return this.FixedAssetName
  }

  /** @returns {string} Tên loại tài sản */
  get categoryName() {
    return this.FixedAssetCategoryName
  }

  /** @returns {string} Tên bộ phận sử dụng */
  get departmentName() {
    return this.DepartmentName
  }

  /** @returns {number} Giá trị hao mòn năm */
  get depreciationValue() {
    return this.DepreciationValue
  }
  // #endregion

  // #region Hàm tiện ích
  /**
   * Chuẩn hóa chuỗi ngày theo định dạng yyyy-MM-dd
   * @param {string|Date} dateStr - Chuỗi ngày hoặc đối tượng Date
   * @returns {string} Ngày đã format
   */
  formatDateString(dateStr) {
    if (!dateStr) return ''
    if (typeof dateStr === 'string') {
      return dateStr.includes('T') ? dateStr.split('T')[0] : dateStr
    }
    const date = new Date(dateStr)
    return date.toISOString().split('T')[0]
  }
  // #endregion

  // #region Chuyển đổi dữ liệu

  /**
   * Chuyển đối tượng FixedAsset sang định dạng dùng cho form chỉnh sửa
   * @returns {Object} Dữ liệu định dạng phục vụ edit popup
   */
  toEditFormat() {
    return {
      FixedAssetId: this.FixedAssetId,
      FixedAssetCode: this.FixedAssetCode,
      FixedAssetName: this.FixedAssetName,

      DepartmentCode: this.DepartmentCode,
      DepartmentName: this.DepartmentName,

      FixedAssetCategoryCode: this.FixedAssetCategoryCode,
      FixedAssetCategoryName: this.FixedAssetCategoryName,

      Quantity: this.Quantity,
      Cost: this.Cost,

      LifeTime: this.LifeTime,
      DepreciationRate: this.DepreciationRate,
      DepreciationValue: this.DepreciationValue,

      PurchaseDate: this.formatDateString(this.PurchaseDate),
      ProductionYear: this.ProductionYear,
      TrackedYear: this.TrackedYear,

      Description: this.Description
    }
  }

  /**
   * Chuyển đối tượng FixedAsset sang định dạng dùng cho bảng (table)
   * @returns {Object} Dữ liệu định dạng phục vụ table
   */
  toTableFormat() {
    return {
      FixedAssetId: this.FixedAssetId,
      CandidateID: this.FixedAssetId,

      FixedAssetCode: this.FixedAssetCode,
      AssetCode: this.FixedAssetCode,

      FixedAssetName: this.FixedAssetName,
      AssetName: this.FixedAssetName,

      DepartmentCode: this.DepartmentCode,
      DepartmentName: this.DepartmentName,

      FixedAssetCategoryCode: this.FixedAssetCategoryCode,
      FixedAssetCategoryName: this.FixedAssetCategoryName,
      AssetTypeName: this.FixedAssetCategoryName,

      Quantity: this.Quantity,
      Cost: this.Cost,

      AccumulatedDepreciation: this.AccumulatedDepreciation,
      RemainingValue: this.RemainingValue,
      DepreciationValue: this.DepreciationValue,

      PurchaseDate: this.PurchaseDate,
      ProductionYear: this.ProductionYear,

      TrackedYear: this.TrackedYear,
      LifeTime: this.LifeTime,
      DepreciationRate: this.DepreciationRate,
    }
  }
  // #endregion

  // #region Static methods
  /**
   * Tạo đối tượng FixedAsset từ dữ liệu API
   * @param {Object} data - Dữ liệu trả về từ API
   * @returns {FixedAsset}
   */
  static fromApi(data) {
    return new FixedAsset(data)
  }

  /**
   * Tạo danh sách FixedAsset từ mảng dữ liệu API
   * @param {Array<Object>} dataArray - Danh sách dữ liệu tài sản
   * @returns {FixedAsset[]}
   */
  static fromApiArray(dataArray) {
    return dataArray.map(data => new FixedAsset(data))
  }
  // #endregion
}

File: MISA.AssetManagement.Api/MISA.AssetManagement.Fresher.Api/MISA.AssetManagement.Core/Entities/FixedAsset.cs
using System;

namespace MISA.Core.Entities
{
    /// <summary>
    /// Thông tin tài sản cố định
    /// CreatedBy: TTVinh (14/11/2025)
    /// </summary>
    public class FixedAsset
    {
        /// <summary>
        /// ID của tài sản cố định
        /// </summary>
        public Guid FixedAssetId { get; set; }

        /// <summary>
        /// Mã tài sản cố định
        /// </summary>
        public string FixedAssetCode { get; set; }

        /// <summary>
        /// Tên tài sản cố định
        /// </summary>
        public string FixedAssetName { get; set; }

        /// <summary>
        /// ID của bộ phận sử dụng
        /// </summary>
        public Guid DepartmentId { get; set; }

        /// <summary>
        /// Mã bộ phận sử dụng
        /// </summary>
        public string DepartmentCode { get; set; }

        /// <summary>
        /// Tên bộ phận sử dụng
        /// </summary>
        public string DepartmentName { get; set; }

        /// <summary>
        /// ID của loại tài sản
        /// </summary>
        public Guid FixedAssetCategoryId { get; set; }

        /// <summary>
        /// Mã loại tài sản
        /// </summary>
        public string FixedAssetCategoryCode { get; set; }

        /// <summary>
        /// Tên loại tài sản
        /// </summary>
        public string FixedAssetCategoryName { get; set; }

        /// <summary>
        /// Ngày mua tài sản
        /// </summary>
        public DateTime PurchaseDate { get; set; }

        /// <summary>
        /// Năm sản xuất của tài sản
        /// </summary>
        public int ProductionYear { get; set; }

        /// <summary>
        /// Năm bắt đầu theo dõi tài sản
        /// </summary>
        public int TrackedYear { get; set; }

        /// <summary>
        /// Thời gian sử dụng (năm)
        /// </summary>
        public int LifeTime { get; set; }

        /// <summary>
        /// Tỷ lệ hao mòn (%)
        /// </summary>
        public decimal DepreciationRate { get; set; }

        /// <summary>
        /// Số lượng tài sản
        /// </summary>
        public int Quantity { get; set; }

        /// <summary>
        /// Nguyên giá của tài sản
        /// </summary>
        public decimal Cost { get; set; }

        /// <summary>
        /// Giá trị hao mòn hàng năm
        /// </summary>
        public decimal DepreciationValue { get; set; }

        /// <summary>
        /// Mô tả thêm về tài sản
        /// </summary>
        public string? Description { get; set; }

        /// <summary>
        /// Trạng thái hoạt động (true: đang hoạt động, false: ngừng sử dụng)
        /// </summary>
        public bool IsActive { get; set; }

        /// <summary>
        /// Ngày tạo bản ghi
        /// </summary>
        public DateTime? CreatedDate { get; set; }

        /// <summary>
        /// Người tạo bản ghi
        /// </summary>
        public string? CreatedBy { get; set; }

        /// <summary>
        /// Ngày chỉnh sửa gần nhất
        /// </summary>
        public DateTime? ModifiedDate { get; set; }

        /// <summary>
        /// Người chỉnh sửa gần nhất
        /// </summary>
        public string? ModifiedBy { get; set; }
    }
}
